File: src/services/invoice.service.ts
/**
 * Servicio de facturación.
 *
 * Regla clave: NO se puede enviar al SRI sin firma electrónica.
 * La firma es un paso obligatorio antes del envío.
 *
 * Estados:
 *   DRAFT → (sign) → SIGNED → (send) → SENT → (check) → AUTHORIZED | REJECTED
 */
import Decimal from "decimal.js";
import { InvoiceStatus, Invoice } from "@prisma/client";
import { z } from "zod";
import { prisma } from "../db/prisma";
import { config } from "../config";
import { generateAccessKey, formatInvoiceNumber } from "../utils/access-key";
import { generateInvoiceXml } from "./sri/xml-generator";
import { signInvoiceXml } from "./sri/xml-signer";
import { getDecryptedP12, getActiveCertificateByRuc } from "./certificate.service";
import { sendToReception, checkAuthorization } from "./sri/sri-client";
import { logger } from "../utils/logger";

// ─── Schemas de entrada ───────────────────────────────────────────
export const CreateInvoiceItemSchema = z.object({
  mainCode: z.string().max(25).optional(),
  auxCode: z.string().max(25).optional(),
  description: z.string().min(1).max(300),
  quantity: z.number().positive(),
  unitPrice: z.number().positive(),
  discount: z.number().min(0).default(0),
  taxPercentageCode: z
    .enum(["0", "2", "3", "6", "7", "10"])
    .default("2"), // 0=0%, 2=12%, 10=15%
});

export const CreateInvoiceSchema = z.object({
  issueDate: z.coerce.date().optional(),
  buyerIdType: z.enum(["04", "05", "06", "07"]).default("05"),
  buyerId: z.string().min(1).max(20),
  buyerName: z.string().min(1).max(300),
  buyerEmail: z.string().email().optional(),
  buyerAddress: z.string().max(300).optional(),
  paymentMethod: z
    .enum(["01", "15", "16", "17", "18", "19", "20", "21"])
    .default("01"),
  items: z.array(CreateInvoiceItemSchema).min(1),
  certificateId: z.string().optional(),
});

export type CreateInvoiceInput = z.infer<typeof CreateInvoiceSchema>;

// ─── Tabla de tarifas IVA ─────────────────────────────────────────
const IVA_RATES: Record<string, number> = {
  "0": 0,    // IVA 0%
  "2": 12,   // IVA 12%
  "3": 14,   // IVA 14%
  "6": 0,    // No objeto de impuesto (no suma)
  "7": 0,    // Exento
  "10": 15,  // IVA 15%
};

// ─── Obtener siguiente secuencial ────────────────────────────────
async function getNextSequential(
  establishment: string,
  emissionPoint: string
): Promise<string> {
  const last = await prisma.invoice.findFirst({
    where: { establishment, emissionPoint },
    orderBy: { sequential: "desc" },
    select: { sequential: true },
  });
  const next = last ? parseInt(last.sequential, 10) + 1 : 1;
  return String(next).padStart(9, "0");
}

// ─── Crear factura (DRAFT) ────────────────────────────────────────
export async function createInvoice(input: CreateInvoiceInput): Promise<Invoice> {
  const validated = CreateInvoiceSchema.parse(input);

  const issueDate = validated.issueDate ?? new Date();
  const establishment = config.ISSUER_ESTABLISHMENT;
  const emissionPoint = config.ISSUER_EMISSION_POINT;
  const sequential = await getNextSequential(establishment, emissionPoint);

  const accessKey = generateAccessKey({
    issueDate,
    ruc: config.ISSUER_RUC,
    environment: config.SRI_ENVIRONMENT,
    establishment,
    emissionPoint,
    sequential,
  });

  // Calcular totales por ítem
  let subtotal = new Decimal(0);
  let totalDiscount = new Decimal(0);
  let totalIva = new Decimal(0);

  const itemsData = validated.items.map((item) => {
    const qty = new Decimal(item.quantity);
    const price = new Decimal(item.unitPrice);
    const disc = new Decimal(item.discount);
    const lineSubtotal = qty.mul(price).minus(disc);
    const taxRate = new Decimal(IVA_RATES[item.taxPercentageCode] ?? 0);
    const taxValue = lineSubtotal.mul(taxRate).div(100);

    subtotal = subtotal.add(lineSubtotal);
    totalDiscount = totalDiscount.add(disc);
    totalIva = totalIva.add(taxValue);

    return {
      mainCode: item.mainCode,
      auxCode: item.auxCode,
      description: item.description,
      quantity: qty.toFixed(6),
      unitPrice: price.toFixed(6),
      discount: disc.toFixed(2),
      subtotalNoTax: lineSubtotal.toFixed(2),
      taxCode: "2",
      taxPercentageCode: item.taxPercentageCode,
      taxRate: taxRate.toFixed(2),
      taxBase: lineSubtotal.toFixed(2),
      taxValue: taxValue.toFixed(2),
    };
  });

  const totalAmount = subtotal.add(totalIva);

  // Determinar certificado
  const certId = validated.certificateId ?? (
    await getActiveCertificateByRuc(config.ISSUER_RUC)
  )?.id ?? null;

  const invoice = await prisma.invoice.create({
    data: {
      accessKey,
      status: "DRAFT",
      environment: config.SRI_ENVIRONMENT,
      issuerRuc: config.ISSUER_RUC,
      issuerName: config.ISSUER_BUSINESS_NAME,
      issuerTradeName: config.ISSUER_TRADE_NAME ?? null,
      issuerAddress: config.ISSUER_ADDRESS,
      establishment,
      emissionPoint,
      sequential,
      issueDate,
      buyerIdType: validated.buyerIdType,
      buyerId: validated.buyerId,
      buyerName: validated.buyerName,
      buyerEmail: validated.buyerEmail ?? null,
      buyerAddress: validated.buyerAddress ?? null,
      subtotal: subtotal.toFixed(2),
      totalDiscount: totalDiscount.toFixed(2),
      totalIva: totalIva.toFixed(2),
      totalAmount: totalAmount.toFixed(2),
      tip: "0.00",
      paymentMethod: validated.paymentMethod,
      paymentAmount: totalAmount.toFixed(2),
      certificateId: certId,
      items: { create: itemsData },
    },
    include: { items: true },
  });

  await prisma.auditLog.create({
    data: {
      invoiceId: invoice.id,
      action: "CREATED",
      details: `Factura ${formatInvoiceNumber(establishment, emissionPoint, sequential)} creada`,
    },
  });

  logger.info("Factura creada", { id: invoice.id, accessKey });
  return invoice;
}

// ─── Firmar factura ───────────────────────────────────────────────
export async function signInvoice(invoiceId: string): Promise<Invoice> {
  const invoice = await prisma.invoice.findUniqueOrThrow({
    where: { id: invoiceId },
    include: { items: true },
  });

  if (invoice.status !== "DRAFT") {
    throw new Error(
      `Solo se pueden firmar facturas en estado DRAFT. Estado actual: ${invoice.status}`
    );
  }

  if (!invoice.certificateId) {
    throw new Error(
      "La factura no tiene un certificado asignado. Suba un certificado P12 antes de firmar."
    );
  }

  // Generar XML
  const xmlContent = generateInvoiceXml(invoice as Parameters<typeof generateInvoiceXml>[0]);

  // Obtener P12 descifrado y firmar
  const { p12Buffer, password } = await getDecryptedP12(invoice.certificateId);
  const signedXml = await signInvoiceXml(xmlContent, p12Buffer, password);

  const updated = await prisma.invoice.update({
    where: { id: invoiceId },
    data: {
      status: "SIGNED",
      xmlContent,
      signedXml,
    },
  });

  await prisma.auditLog.create({
    data: {
      invoiceId,
      action: "SIGNED",
      details: "XML firmado con XAdES-BES",
    },
  });

  logger.info("Factura firmada", { id: invoiceId });
  return updated;
}

// ─── Enviar al SRI (solo si está firmada) ────────────────────────
export async function sendInvoiceToSri(invoiceId: string): Promise<Invoice> {
  const invoice = await prisma.invoice.findUniqueOrThrow({
    where: { id: invoiceId },
  });

  if (invoice.status !== "SIGNED") {
    throw new Error(
      `No se puede enviar al SRI una factura sin firma. Estado: ${invoice.status}. ` +
        "Firme primero la factura."
    );
  }

  if (!invoice.signedXml) {
    throw new Error("El XML firmado no está disponible. Firme la factura primero.");
  }

  const reception = await sendToReception(invoice.signedXml);

  let updated: Invoice;

  if (reception.state === "RECIBIDA") {
    updated = await prisma.invoice.update({
      where: { id: invoiceId },
      data: {
        status: "SENT",
        sriResponse: JSON.stringify(reception),
      },
    });
    await prisma.auditLog.create({
      data: {
        invoiceId,
        action: "SENT",
        details: `Recibida por SRI. Mensajes: ${JSON.stringify(reception.messages)}`,
      },
    });
    logger.info("Factura enviada al SRI", { id: invoiceId });
  } else {
    const reason = reception.messages.map((m) => m.message).join("; ");
    updated = await prisma.invoice.update({
      where: { id: invoiceId },
      data: {
        status: "REJECTED",
        sriResponse: JSON.stringify(reception),
        rejectReason: reason,
      },
    });
    await prisma.auditLog.create({
      data: {
        invoiceId,
        action: "REJECTED",
        details: `SRI rechazó la recepción: ${reason}`,
      },
    });
    logger.warn("SRI rechazó la factura en recepción", { id: invoiceId, reason });
  }

  return updated;
}

// ─── Verificar autorización ───────────────────────────────────────
export async function checkInvoiceAuthorization(invoiceId: string): Promise<Invoice> {
  const invoice = await prisma.invoice.findUniqueOrThrow({
    where: { id: invoiceId },
  });

  if (invoice.status !== "SENT") {
    throw new Error(
      `Solo se puede verificar autorización de facturas en estado SENT. Estado: ${invoice.status}`
    );
  }

  const auth = await checkAuthorization(invoice.accessKey);

  let updated: Invoice;

  if (auth.state === "AUTORIZADO") {
    updated = await prisma.invoice.update({
      where: { id: invoiceId },
      data: {
        status: "AUTHORIZED",
        authNumber: auth.authNumber,
        authDate: auth.authDate ? new Date(auth.authDate) : null,
        authXml: auth.comprobante,
      },
    });
    await prisma.auditLog.create({
      data: {
        invoiceId,
        action: "AUTHORIZED",
        details: `Autorizada por SRI. Número: ${auth.authNumber}`,
      },
    });
    logger.info("Factura autorizada por SRI", {
      id: invoiceId,
      authNumber: auth.authNumber,
    });
  } else {
    const reason = auth.messages.map((m) => m.message).join("; ");
    updated = await prisma.invoice.update({
      where: { id: invoiceId },
      data: {
        status: "REJECTED",
        rejectReason: reason,
      },
    });
    await prisma.auditLog.create({
      data: {
        invoiceId,
        action: "REJECTED",
        details: `SRI no autorizó: ${reason}`,
      },
    });
    logger.warn("Factura no autorizada por SRI", { id: invoiceId, reason });
  }

  return updated;
}

// ─── Cancelar factura (solo DRAFT o SIGNED) ───────────────────────
export async function cancelInvoice(invoiceId: string, reason?: string): Promise<Invoice> {
  const invoice = await prisma.invoice.findUniqueOrThrow({
    where: { id: invoiceId },
  });

  if (!["DRAFT", "SIGNED"].includes(invoice.status)) {
    throw new Error(
      `Solo se pueden cancelar facturas en estado DRAFT o SIGNED. Estado: ${invoice.status}`
    );
  }

  const updated = await prisma.invoice.update({
    where: { id: invoiceId },
    data: { status: "CANCELLED" },
  });

  await prisma.auditLog.create({
    data: {
      invoiceId,
      action: "CANCELLED",
      details: reason ?? "Cancelada manualmente",
    },
  });

  return updated;
}

// ─── Consultas ────────────────────────────────────────────────────
export async function getInvoice(id: string) {
  return prisma.invoice.findUniqueOrThrow({
    where: { id },
    include: { items: true, auditLogs: { orderBy: { createdAt: "asc" } } },
  });
}

export async function listInvoices(filters: {
  status?: InvoiceStatus;
  page?: number;
  limit?: number;
}) {
  const page = filters.page ?? 1;
  const limit = Math.min(filters.limit ?? 20, 100);
  const skip = (page - 1) * limit;

  const [total, invoices] = await Promise.all([
    prisma.invoice.count({ where: { status: filters.status } }),
    prisma.invoice.findMany({
      where: { status: filters.status },
      orderBy: { createdAt: "desc" },
      skip,
      take: limit,
      include: { items: true },
    }),
  ]);

  return { total, page, limit, invoices };
}

File: src/VenturacomSri.Api/Services/InvoiceService.cs
using Microsoft.EntityFrameworkCore;
using VenturacomSri.Api.Data;
using VenturacomSri.Api.DTOs;
using VenturacomSri.Api.Models;
using VenturacomSri.Api.Services.Sri;
using VenturacomSri.Api.Utils;

namespace VenturacomSri.Api.Services;

/// <summary>
/// Servicio de facturación.
///
/// Regla de seguridad: NO se puede enviar al SRI sin firma digital.
/// El método <see cref="SendToSriAsync"/> lanza si la factura no está firmada.
/// </summary>
public class InvoiceService
{
    private readonly AppDbContext _db;
    private readonly IConfiguration _cfg;
    private readonly XmlGeneratorService _xmlGen;
    private readonly XmlSignerService _xmlSigner;
    private readonly CertificateService _certService;
    private readonly SriClientService _sriClient;
    private readonly ILogger<InvoiceService> _logger;

    // IVA por código de porcentaje
    private static readonly Dictionary<string, decimal> IvaTariff = new()
    {
        ["0"]  = 0m,
        ["2"]  = 12m,
        ["3"]  = 14m,
        ["6"]  = 0m,
        ["7"]  = 0m,
        ["10"] = 15m,
    };

    public InvoiceService(
        AppDbContext db,
        IConfiguration cfg,
        XmlGeneratorService xmlGen,
        XmlSignerService xmlSigner,
        CertificateService certService,
        SriClientService sriClient,
        ILogger<InvoiceService> logger)
    {
        _db          = db;
        _cfg         = cfg;
        _xmlGen      = xmlGen;
        _xmlSigner   = xmlSigner;
        _certService = certService;
        _sriClient   = sriClient;
        _logger      = logger;
    }

    // ── Crear factura (estado DRAFT) ──────────────────────────────
    public async Task<Invoice> CreateAsync(CreateInvoiceRequest req)
    {
        var issuerRuc = _cfg["Issuer:Ruc"]!;
        var env       = int.Parse(_cfg["Sri:Environment"] ?? "1");
        var estab     = _cfg["Issuer:Establishment"] ?? "001";
        var ptoEmi    = _cfg["Issuer:EmissionPoint"] ?? "001";
        var issueDate = req.IssueDate ?? DateTime.UtcNow;

        var sequential = await GetNextSequentialAsync(estab, ptoEmi);
        var accessKey  = AccessKeyGenerator.Generate(
            issueDate, issuerRuc, env, estab, ptoEmi, sequential);

        // Calcular totales
        decimal subtotal = 0, totalDiscount = 0, totalIva = 0;
        var itemsData = new List<InvoiceItem>();

        foreach (var itemReq in req.Items)
        {
            var qty      = itemReq.Quantity;
            var price    = itemReq.UnitPrice;
            var disc     = itemReq.Discount;
            var lineBase = qty * price - disc;
            var rate     = IvaTariff.GetValueOrDefault(itemReq.TaxPercentageCode, 0m);
            var taxVal   = Math.Round(lineBase * rate / 100, 2);

            subtotal      += lineBase;
            totalDiscount += disc;
            totalIva      += taxVal;

            itemsData.Add(new InvoiceItem
            {
                MainCode         = itemReq.MainCode,
                AuxCode          = itemReq.AuxCode,
                Description      = itemReq.Description,
                Quantity         = qty,
                UnitPrice        = price,
                Discount         = disc,
                SubtotalNoTax    = lineBase,
                TaxCode          = "2",
                TaxPercentageCode = itemReq.TaxPercentageCode,
                TaxRate          = rate,
                TaxBase          = lineBase,
                TaxValue         = taxVal,
            });
        }

        var total = subtotal + totalIva;

        // Buscar certificado activo si no se especificó
        var certId = req.CertificateId
            ?? (await _certService.GetActiveByRucAsync(issuerRuc))?.Id;

        var invoice = new Invoice
        {
            AccessKey       = accessKey,
            Status          = InvoiceStatus.Draft,
            Environment     = env,
            IssuerRuc       = issuerRuc,
            IssuerName      = _cfg["Issuer:BusinessName"]!,
            IssuerTradeName = _cfg["Issuer:TradeName"],
            IssuerAddress   = _cfg["Issuer:Address"]!,
            Establishment   = estab,
            EmissionPoint   = ptoEmi,
            Sequential      = sequential,
            IssueDate       = issueDate,
            BuyerIdType     = req.BuyerIdType,
            BuyerId         = req.BuyerId,
            BuyerName       = req.BuyerName,
            BuyerEmail      = req.BuyerEmail,
            BuyerAddress    = req.BuyerAddress,
            Subtotal        = Math.Round(subtotal, 2),
            TotalDiscount   = Math.Round(totalDiscount, 2),
            TotalIva        = Math.Round(totalIva, 2),
            TotalAmount     = Math.Round(total, 2),
            PaymentMethod   = req.PaymentMethod,
            PaymentAmount   = Math.Round(total, 2),
            CertificateId   = certId,
        };

        invoice.Items = itemsData;
        _db.Invoices.Add(invoice);
        await _db.SaveChangesAsync();

        await AddAuditAsync(invoice.Id, "CREATED",
            $"Factura {AccessKeyGenerator.FormatNumber(estab, ptoEmi, sequential)} creada");

        _logger.LogInformation("Factura creada {Id} clave:{Key}", invoice.Id, accessKey);
        return invoice;
    }

    // ── Firmar factura ────────────────────────────────────────────
    public async Task<Invoice> SignAsync(string invoiceId)
    {
        var invoice = await LoadWithItemsAsync(invoiceId);

        if (invoice.Status != InvoiceStatus.Draft)
            throw new InvalidOperationException(
                $"Solo se pueden firmar facturas DRAFT. Estado actual: {invoice.Status}.");

        if (string.IsNullOrEmpty(invoice.CertificateId))
            throw new InvalidOperationException(
                "La factura no tiene certificado asignado. " +
                "Suba un certificado P12 y asígnelo antes de firmar.");

        // Generar XML sin firma
        var xml = _xmlGen.GenerateInvoiceXml(invoice);

        // Obtener P12 descifrado y firmar
        var (p12, pwd) = await _certService.GetDecryptedAsync(invoice.CertificateId);
        var cert509    = _xmlSigner.LoadCertificate(p12, pwd);
        var signedXml  = _xmlSigner.SignXml(xml, cert509);

        invoice.Status     = InvoiceStatus.Signed;
        invoice.XmlContent = xml;
        invoice.SignedXml  = signedXml;
        invoice.UpdatedAt  = DateTime.UtcNow;
        await _db.SaveChangesAsync();

        await AddAuditAsync(invoiceId, "SIGNED", "XML firmado con XAdES-BES");
        _logger.LogInformation("Factura firmada {Id}", invoiceId);
        return invoice;
    }

    // ── Enviar al SRI (solo facturas FIRMADAS) ────────────────────
    public async Task<Invoice> SendToSriAsync(string invoiceId, CancellationToken ct = default)
    {
        var invoice = await _db.Invoices.FindAsync(new object[] { invoiceId }, ct)
            ?? throw new KeyNotFoundException(invoiceId);

        if (invoice.Status != InvoiceStatus.Signed)
            throw new InvalidOperationException(
                $"No se puede enviar al SRI una factura sin firma electrónica. " +
                $"Estado: {invoice.Status}. Firme primero la factura con POST /invoices/{invoiceId}/sign.");

        if (string.IsNullOrEmpty(invoice.SignedXml))
            throw new InvalidOperationException(
                "El XML firmado no está disponible. Firme la factura primero.");

        var result = await _sriClient.SendToReceptionAsync(invoice.SignedXml, ct);

        if (result.State == ReceptionState.Received)
        {
            invoice.Status      = InvoiceStatus.Sent;
            invoice.SriResponse = System.Text.Json.JsonSerializer.Serialize(result);
            invoice.UpdatedAt   = DateTime.UtcNow;
            await _db.SaveChangesAsync();

            await AddAuditAsync(invoiceId, "SENT",
                $"Recibida por SRI. Mensajes: {string.Join("; ", result.Messages.Select(m => m.Message))}");
        }
        else
        {
            var reason = string.Join("; ", result.Messages.Select(m => m.Message));
            invoice.Status      = InvoiceStatus.Rejected;
            invoice.RejectReason = reason;
            invoice.SriResponse  = System.Text.Json.JsonSerializer.Serialize(result);
            invoice.UpdatedAt    = DateTime.UtcNow;
            await _db.SaveChangesAsync();

            await AddAuditAsync(invoiceId, "REJECTED", $"SRI devolvió: {reason}");
            _logger.LogWarning("SRI rechazó la factura {Id}: {Reason}", invoiceId, reason);
        }

        return invoice;
    }

    // ── Verificar autorización ────────────────────────────────────
    public async Task<Invoice> CheckAuthorizationAsync(string invoiceId, CancellationToken ct = default)
    {
        var invoice = await _db.Invoices.FindAsync(new object[] { invoiceId }, ct)
            ?? throw new KeyNotFoundException(invoiceId);

        if (invoice.Status != InvoiceStatus.Sent)
            throw new InvalidOperationException(
                $"Solo se puede verificar la autorización de facturas SENT. Estado: {invoice.Status}.");

        var auth = await _sriClient.CheckAuthorizationAsync(invoice.AccessKey, ct);

        if (auth.State == AuthorizationState.Authorized)
        {
            invoice.Status     = InvoiceStatus.Authorized;
            invoice.AuthNumber = auth.AuthNumber;
            invoice.AuthDate   = DateTime.TryParse(auth.AuthDate, out var d) ? d.ToUniversalTime() : null;
            invoice.AuthXml    = auth.Comprobante;
            invoice.UpdatedAt  = DateTime.UtcNow;
            await _db.SaveChangesAsync();

            await AddAuditAsync(invoiceId, "AUTHORIZED",
                $"Autorizada SRI. Número: {auth.AuthNumber}");
            _logger.LogInformation("Factura autorizada {Id} auth:{Num}", invoiceId, auth.AuthNumber);
        }
        else
        {
            var reason = string.Join("; ", auth.Messages.Select(m => m.Message));
            invoice.Status       = InvoiceStatus.Rejected;
            invoice.RejectReason = reason;
            invoice.UpdatedAt    = DateTime.UtcNow;
            await _db.SaveChangesAsync();

            await AddAuditAsync(invoiceId, "REJECTED", $"SRI no autorizó: {reason}");
            _logger.LogWarning("Factura no autorizada {Id}: {Reason}", invoiceId, reason);
        }

        return invoice;
    }

    // ── Cancelar (solo DRAFT o SIGNED) ────────────────────────────
    public async Task<Invoice> CancelAsync(string invoiceId, string? reason = null)
    {
        var invoice = await _db.Invoices.FindAsync(invoiceId)
            ?? throw new KeyNotFoundException(invoiceId);

        if (invoice.Status is not (InvoiceStatus.Draft or InvoiceStatus.Signed))
            throw new InvalidOperationException(
                $"Solo se pueden cancelar facturas DRAFT o SIGNED. Estado: {invoice.Status}.");

        invoice.Status    = InvoiceStatus.Cancelled;
        invoice.UpdatedAt = DateTime.UtcNow;
        await _db.SaveChangesAsync();
        await AddAuditAsync(invoiceId, "CANCELLED", reason ?? "Cancelada manualmente");
        return invoice;
    }

    // ── Consultas ─────────────────────────────────────────────────
    public async Task<Invoice> GetAsync(string id)
        => await _db.Invoices
            .Include(i => i.Items)
            .Include(i => i.AuditLogs.OrderBy(a => a.CreatedAt))
            .FirstOrDefaultAsync(i => i.Id == id)
        ?? throw new KeyNotFoundException(id);

    public async Task<(int Total, List<Invoice> Items)> ListAsync(
        InvoiceStatus? status, int page, int limit)
    {
        var query = _db.Invoices
            .Include(i => i.Items)
            .AsQueryable();
        if (status.HasValue) query = query.Where(i => i.Status == status.Value);

        var total = await query.CountAsync();
        var items = await query
            .OrderByDescending(i => i.CreatedAt)
            .Skip((page - 1) * limit)
            .Take(limit)
            .ToListAsync();

        return (total, items);
    }

    // ── Helpers ───────────────────────────────────────────────────
    private async Task<Invoice> LoadWithItemsAsync(string id)
        => await _db.Invoices
            .Include(i => i.Items)
            .FirstOrDefaultAsync(i => i.Id == id)
        ?? throw new KeyNotFoundException(id);

    private async Task<string> GetNextSequentialAsync(string estab, string ptoEmi)
    {
        var last = await _db.Invoices
            .Where(i => i.Establishment == estab && i.EmissionPoint == ptoEmi)
            .OrderByDescending(i => i.Sequential)
            .Select(i => i.Sequential)
            .FirstOrDefaultAsync();

        var next = last is null ? 1 : int.Parse(last) + 1;
        return next.ToString().PadLeft(9, '0');
    }

    private async Task AddAuditAsync(string invoiceId, string action, string? details = null)
    {
        _db.AuditLogs.Add(new AuditLog
        {
            InvoiceId = invoiceId,
            Action    = action,
            Details   = details,
        });
        await _db.SaveChangesAsync();
    }
}
